include/gsl/span

```diff
@@ -451,62 +494,74 @@ public:
-        Expects(Count >= 0 && Count <= size());
+        Expects(Count != dynamic_extent);
+        Expects(Count <= size());
         return {data(), Count};
     }
 
-    template <std::ptrdiff_t Count>
+    template <std::size_t Count>
+    // clang-format off
     GSL_SUPPRESS(bounds.1) // NO-FORMAT: attribute
-    constexpr span<element_type, Count> last() const noexcept
+        // clang-format on
+        constexpr span<element_type, Count> last() const noexcept
     {
-        Expects(Count >= 0 && size() - Count >= 0);
+        Expects(Count != dynamic_extent);
+        Expects(size() >= Count);
         return {data() + (size() - Count), Count};
     }
 
-    template <std::ptrdiff_t Offset, std::ptrdiff_t Count = dynamic_extent>
+    template <std::size_t Offset, std::size_t Count = dynamic_extent>
+    // clang-format off
     GSL_SUPPRESS(bounds.1) // NO-FORMAT: attribute
-    constexpr auto subspan() const noexcept ->
+        // clang-format on
+        constexpr auto subspan() const noexcept ->
         typename details::calculate_subspan_type<ElementType, Extent, Offset, Count>::type
     {
-        Expects((Offset >= 0 && size() - Offset >= 0) &&
-                (Count == dynamic_extent || (Count >= 0 && Offset + Count <= size())));
+        Expects((size() >= Offset) && (Count == dynamic_extent || (Count <= size() - Offset)));
 
         return {data() + Offset, Count == dynamic_extent ? size() - Offset : Count};
     }
 
-    constexpr span<element_type, dynamic_extent> first(index_type count) const noexcept
+    constexpr span<element_type, dynamic_extent> first(size_type count) const noexcept
     {
-        Expects(count >= 0 && count <= size());
+        Expects(count <= size());
         return {data(), count};
     }
 
-    constexpr span<element_type, dynamic_extent> last(index_type count) const noexcept
+    constexpr span<element_type, dynamic_extent> last(size_type count) const noexcept
     {
+        Expects(count <= size());
         return make_subspan(size() - count, dynamic_extent, subspan_selector<Extent>{});
     }
 
-    constexpr span<element_type, dynamic_extent> subspan(index_type offset,
-                                                         index_type count = dynamic_extent) const noexcept
+    constexpr span<element_type, dynamic_extent> subspan(size_type offset,
+                                                         size_type count = dynamic_extent) const
+        noexcept
     {
         return make_subspan(offset, count, subspan_selector<Extent>{});
     }
 
     // [span.obs], span observers
-    constexpr index_type size() const noexcept { return storage_.size(); }
-    constexpr index_type size_bytes() const noexcept
+    constexpr size_type size() const noexcept { return storage_.size(); }
+
+    constexpr size_type size_bytes() const noexcept
     {
-        return size() * narrow_cast<index_type>(sizeof(element_type));
+        Expects(size() < dynamic_extent / sizeof(element_type));
```

Comment:
If this Expects is an invariant it seems like it should be on the constructors rather than here? After all, if this condition ever happens that the span isn't meeting the "is a range of N objects of type T" invariant.